File: training-data/matlab/sliding_window_max.m
function result = sliding_window_max(nums, k)
    n = numel(nums);
    result = zeros(1, n - k + 1);
    deque = [];
    for i = 1:n
        while ~isempty(deque) && deque(1) <= i - k
            deque(1) = [];
        end
        while ~isempty(deque) && nums(deque(end)) <= nums(i)
            deque(end) = [];
        end
        deque(end + 1) = i;
        if i >= k
            result(i - k + 1) = nums(deque(1));
        end
    end
end

nums = [1, 3, -1, -3, 5, 3, 6, 7];
disp(sliding_window_max(nums, 3))

File: training-data/r/sliding_window_max.R
sliding_window_max <- function(nums, k) {
  result <- c()
  deque <- c()

  for (i in seq_along(nums)) {
    while (length(deque) > 0 && deque[1] <= i - k) {
      deque <- deque[-1]
    }
    while (length(deque) > 0 && nums[deque[length(deque)]] <= nums[i]) {
      deque <- deque[-length(deque)]
    }
    deque <- c(deque, i)

    if (i >= k) {
      result <- c(result, nums[deque[1]])
    }
  }

  result
}

nums <- c(1, 3, -1, -3, 5, 3, 6, 7)
print(sliding_window_max(nums, 3))
